src/snapshot/component_checksum_hash.rs

```diff
@@ -0,0 +1,64 @@
+use std::{
+    collections::hash_map::DefaultHasher,
+    hash::{Hash, Hasher},
+    marker::PhantomData,
+};
+
+use bevy::prelude::*;
+
+use crate::{ChecksumFlag, ChecksumPart, Rollback, SaveWorld, SaveWorldSet};
+
+/// A [`Plugin`] which will track the [`Component`] `C` on [`Rollback Entities`](`Rollback`) and ensure a
+/// [`ChecksumPart`] is available and updated. This can be used to generate a [`Checksum`](`crate::Checksum`).
+pub struct GgrsComponentChecksumHashPlugin<C>
+where
+    C: Component + Hash,
+{
+    _phantom: PhantomData<C>,
+}
+
+impl<C> Default for GgrsComponentChecksumHashPlugin<C>
+where
+    C: Component + Hash,
+{
+    fn default() -> Self {
+        Self {
+            _phantom: Default::default(),
+        }
+    }
+}
+
+impl<C> GgrsComponentChecksumHashPlugin<C>
+where
+    C: Component + Hash,
+{
+    /// A [`System`] responsible for managing a [`ChecksumPart`] for the [`Component`] type `C`.
+    pub fn update(
+        mut commands: Commands,
+        components: Query<&C, (With<Rollback>, Without<ChecksumFlag<C>>)>,
+        mut checksum: Query<&mut ChecksumPart, (Without<Rollback>, With<ChecksumFlag<C>>)>,
+    ) {
+        let mut hasher = DefaultHasher::new();
+
+        for component in components.iter() {
+            component.hash(&mut hasher);
+        }
+
+        let result = ChecksumPart(hasher.finish());
+
+        if let Ok(mut checksum) = checksum.get_single_mut() {
+            *checksum = result;
+        } else {
+            commands.spawn((result, ChecksumFlag::<C>::default()));
+        }
+    }
+}
+
+impl<C> Plugin for GgrsComponentChecksumHashPlugin<C>
+where
+    C: Component + Hash,
+{
+    fn build(&self, app: &mut App) {
+        app.add_systems(SaveWorld, Self::update.in_set(SaveWorldSet::Snapshot));
```

Comment:
Would it make sense with a checksum set (honest question don't know the code well enough yet)?

Reply:
I don't think so. The way I see it, the `SaveWorldSet::Snapshot` should be an entirely read-only set for anything affecting the world. It would be trivial to add an extra set for it, but I don't think it would make a semantic difference.